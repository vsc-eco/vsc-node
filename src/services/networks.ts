export default {
    //LEGACY DONT USE
    "testnet/ab8b6cf1-b344-4ad3-8f81-f2d72c61f6b2": {
        genesisDay: 70790692,
        roundLength: 20,
        totalRounds: 60,
        //10 Rounds per every consensus period. 
        //Thus 20 blocks per round * 3 seconds = 60 * 10 = 600 seconds = 10 minutes
        consensusRoundLength: 10, 
        multisigAccount: "null"
    },
    //TESTNET v2
    "testnet/d12e6110-9c8c-4498-88f8-67ddf90d451c": {
        genesisDay: 74869131,
        roundLength: 10, //30 seconds
        totalRounds: 120,
        //10 Rounds per every consensus period. 
        //Thus 20 blocks per round * 3 seconds = 60 * 10 = 600 seconds = 10 minutes
        consensusRoundLength: 10, 
        multisigAccount: 'vsc.ptk-d12e6110'
    },
    //HiveFest dedicated
    'testnet/0bf2e474-6b9e-4165-ad4e-a0d78968d20c': {
        genesisDay: 78559454,
        roundLength: 5, //15 seconds
        totalRounds: 240,
        //10 Rounds per every consensus period. 
        //Thus 20 blocks per round * 3 seconds = 60 * 10 = 600 seconds = 10 minutes
        consensusRoundLength: 10,
        multisigAccount: 'vsc.ptk-d12e6110'
    }
} as Record<string, {
    genesisDay: number,
    roundLength: number
    totalRounds: number
    multisigAccount?: string
}>