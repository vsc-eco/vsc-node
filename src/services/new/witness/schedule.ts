import networks from '../../../services/networks'
import seedrandom from 'seedrandom'
import type { WithId } from 'mongodb'
import type { ElectionResult } from './electionManager'
import type { NewCoreService } from '..'
import type { WitnessServiceV2 } from '.'

function seedRand(func: () => number, min: number, max: number): number {
  return Math.floor(func() * (max - min + 1)) + min
}

function shuffle<T>(arr: T[], seed: string): T[] {
  const size = arr.length
  const rng = seedrandom(seed)
  const resp: T[] = []
  const keys: number[] = []

  for (let i = 0; i < size; i++) keys.push(i)
  for (let i = 0; i < size; i++) {
    const r = seedRand(rng, 0, keys.length - 1)
    const g = keys[r]
    keys.splice(r, 1)
    resp.push(arr[g])
  }
  return resp
}

type WithScheduleBN<T> = T & {
  bn: number
  bn_works: boolean
  in_past: boolean
}

/**
 * Applies block numbers to witness schedule
 */
function applyBNSchedule<T>(
  schedule: T[],
  network: NetworkInfo,
  consensusRound: ConsensusRound,
): WithScheduleBN<T>[] {
  const { roundLength } = network

  return schedule.map((e, index) => {
    return {
      ...e,
      bn: consensusRound.pastRoundHeight + index * roundLength,
      bn_works: (consensusRound.pastRoundHeight + index * roundLength) % roundLength === 0,
      in_past:
        consensusRound.pastRoundHeight + index * roundLength < consensusRound.currentBlockNumber,
    }
  })
}

function weightedSchedule(
  network: NetworkInfo,
  witnessNodes: {
    account: string
    key: string
  }[],
  electionResult: WithId<ElectionResult>,
  randomizeHash: string,
  consensusRound: ConsensusRound,
) {
  const { totalRounds } = network

  const outSchedule: {
    account: string
    key: string
  }[] = []
  for (let x = 0; x < totalRounds; x++) {
    if (witnessNodes[x % witnessNodes.length]) {
      outSchedule.push(witnessNodes[x % witnessNodes.length])
    }
  }

  const schedule = applyBNSchedule(shuffle(outSchedule, randomizeHash), network, consensusRound)

  return {
    schedule,
    valid_from: schedule[0]?.bn || 0,
    valid_to: schedule[outSchedule.length - 1]?.bn || 1,
    valid_epoch: electionResult.epoch,
  }
}

// networks[this.self.config.get('network.id')]
type NetworkInfo = typeof networks[keyof typeof networks]

type ConsensusRound = {
  nextRoundHeight: number
  pastRoundHeight: number
  currentBlockNumber: number
}

export function calculateConsensusRound(blockNumber: number, network: NetworkInfo): ConsensusRound {
  const { roundLength, totalRounds } = network

  const modLength = roundLength * totalRounds
  const mod3 = blockNumber % modLength
  const pastRoundHeight = blockNumber - mod3

  return {
    nextRoundHeight: blockNumber + (modLength - mod3),
    pastRoundHeight,
    currentBlockNumber: blockNumber,
  }
}

async function getRandomizeHashForBlock(
  { self }: { self: NewCoreService },
  consensusRound: ConsensusRound,
): Promise<string> {
  const { pastRoundHeight } = consensusRound

  //Return block id of previous block to round start.
  //This is used for schedule randomization
  const blockHeader = await self.chainBridge.events.findOne({
    key: pastRoundHeight - 1,
  })

  if (!blockHeader) {
    throw new Error(`could not find hive block header for block: ${pastRoundHeight - 1}`)
  }

  return blockHeader.block_id
}

export type Schedule = ReturnType<typeof weightedSchedule>['schedule']

/**
 * Get block producer schedule
 * @param blockHeight
 * @returns
 */
export async function getBlockSchedule(witness: WitnessServiceV2, blockHeight: number): Promise<Schedule> {
  const { self } = witness
  const network: NetworkInfo = networks[self.config.get('network.id')]
  const consensusRound = calculateConsensusRound(blockHeight, network)
  const electionResult = await self.electionManager.getValidElectionOfblock(blockHeight)
  // console.log(electionResult.epoch, this.witnessSchedule.valid_epoch, consensusRound.randomizeHash, this.witnessSchedule.valid_height)
  if (
    witness.witnessSchedule &&
    witness.witnessSchedule.valid_height === consensusRound.pastRoundHeight &&
    electionResult?.epoch === witness.witnessSchedule.valid_epoch
  ) {
    // console.log('this.witnessSchedule.valid_to', this.witnessSchedule.valid_height, blockHeight, consensusRound)
    return witness.witnessSchedule.schedule
  }

  const [witnessNodes, randomizeHash] = await Promise.all([self.electionManager.getMembersOfBlock(blockHeight),getRandomizeHashForBlock(witness, consensusRound)])

  const { schedule, valid_to, valid_from, valid_epoch } = weightedSchedule(
    network,
    witnessNodes,
    electionResult,
    randomizeHash,
    consensusRound,
  )

  witness.witnessSchedule = {
    schedule,
    valid_epoch,
    valid_height: consensusRound.pastRoundHeight,
  }

  return schedule
}
