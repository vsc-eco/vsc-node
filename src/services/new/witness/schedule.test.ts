import networks from '../../networks'
import { calculateConsensusRound, weightedSchedule } from './schedule'

const network = networks['testnet/0bf2e474-6b9e-4165-ad4e-a0d78968d20c']

describe('schedule', () => {
    describe('calculateConsensusRound', () => {
        it('should work in between rounds', () => {
            const block = 10
            const round = calculateConsensusRound(block, network)
            expect(round.currentBlockNumber).toBe(block)
            expect(round.nextRoundHeight).toBe(1200)
            expect(round.pastRoundHeight).toBe(0)
        })
        it('should work at the start of rounds', () => {
            const block = 0
            const round = calculateConsensusRound(block, network)
            expect(round.currentBlockNumber).toBe(block)
            expect(round.nextRoundHeight).toBe(1200)
            expect(round.pastRoundHeight).toBe(0)
        })
        it('should work at the end of rounds', () => {
            const block = 1200
            const round = calculateConsensusRound(block, network)
            expect(round.currentBlockNumber).toBe(block)
            expect(round.nextRoundHeight).toBe(2400)
            expect(round.pastRoundHeight).toBe(1200)
        })
    })

    describe('weightedSchedule', () => {
        type Params = Parameters<typeof weightedSchedule>;
        type WitnessNodes = Params[1];
        type RandomizeHash = Params[2];

        type Return = ReturnType<typeof weightedSchedule>;
        type Schedule = Return['schedule'];

        const account = (name: string): WitnessNodes[number] => ({
          account: name,
          key: 'redacted',
        })

        const tests: {
          it: string
          inputs: {
            block: number
            witnessNodes: WitnessNodes
            randomizeHash: RandomizeHash
          }
          outputs: {
            schedule: Schedule
          }
        }[] = [
          {
            it: 'should work',
            inputs: {
              // 0, 1, ..., 8, 9, 10, 11
              // 0, 10, 20, 30, ...
              block: 8, // in_past === true, for all bn < block
              // missing some or all elections
              witnessNodes: [account('A'), account('B'), account('C')], // which data is getting randomized
              // missing hive blocks or bad data
              randomizeHash: '', // the rng seed used to randomize ^
            },
            outputs: {
              schedule: [], // using snapshot since array is too long for this file
            },
          },
        ]

        for (const test of tests) {
            it(test.it, () => {
                const {block, witnessNodes, randomizeHash} = test.inputs
                const round = calculateConsensusRound(block, network)
                const {schedule} = weightedSchedule(network, witnessNodes, randomizeHash, round)
                expect(schedule).toMatchSnapshot()
            })
        }
    })
})