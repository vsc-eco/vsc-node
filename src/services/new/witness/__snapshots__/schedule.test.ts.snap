// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schedule weightedSchedule should work 1`] = `
[
  {
    "account": "A",
    "bn": 0,
    "bn_works": true,
    "in_past": true,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 10,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 20,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 30,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 40,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 50,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 60,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 70,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 80,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 90,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 100,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 110,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 120,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 130,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 140,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 150,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 160,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 170,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 180,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 190,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 200,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 210,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 220,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 230,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 240,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 250,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 260,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 270,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 280,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 290,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 300,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 310,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 320,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 330,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 340,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 350,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 360,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 370,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 380,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 390,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 400,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 410,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 420,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 430,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 440,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 450,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 460,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 470,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 480,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 490,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 500,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 510,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 520,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 530,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 540,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 550,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 560,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 570,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 580,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 590,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 600,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 610,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 620,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 630,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 640,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 650,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 660,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 670,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 680,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 690,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 700,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 710,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 720,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 730,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 740,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 750,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 760,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 770,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 780,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 790,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 800,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 810,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 820,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 830,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 840,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 850,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 860,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 870,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 880,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 890,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 900,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 910,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 920,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 930,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 940,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 950,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 960,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 970,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 980,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 990,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 1000,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 1010,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 1020,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 1030,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 1040,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 1050,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 1060,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 1070,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 1080,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 1090,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 1100,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 1110,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 1120,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 1130,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 1140,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 1150,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 1160,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "C",
    "bn": 1170,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "A",
    "bn": 1180,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
  {
    "account": "B",
    "bn": 1190,
    "bn_works": true,
    "in_past": false,
    "key": "redacted",
  },
]
`;
